Add support for embedded bots

robot-dreams opened the pull request:
This PR gives us a way to run bots directly from the Zulip server.  This uses a very similar approach to @vabs22 's outgoing webhook system in #4741, but instead of making a REST call in the `QueueProcessingWorker`, it delegates to the bot handler.

I've tested this locally with the `xkcd` bot; I've also added a few unit tests to ensure that events get enqueued as expected.

The process for adding an embedded bot is as follows:
- Create a bot user with `bot_type == UserProfile.EMBEDDED_BOT`
- Create a `Service` object with the correct value for the `name` and `user_profile_id` fields (see the comments in `models.py` below for more details)

Once this is merged, there are a lot of additional improvements we can make:
- [ ] Add management scripts / UI functionality so that it's really easy to turn on an embedded bot
- [ ] Add a whitelist for approved embedded bots
- [ ] Refactor the bot handler code so that bots return the response (instead of sending a message directly)
- [ ] Don't use `sys.exit(1)` in `bot_lib.py` (we shouldn't deploy any embedded bots until this is fixed)
- [ ] Clean up the at-mention handling in `bot_lib.py` and consolidate it with the code in the `QueueProcessingWorker`
- [ ] Handle stateful embedded bots
- [ ] If necessary, add caching for bot clients / handlers

smarx commented:
*Automated message from Dropbox CLA bot*

@robot-dreams, it looks like you've already signed the Dropbox CLA. Thanks!

robot-dreams commented on zerver/lib/actions.py:
```diff
@@ -773,10 +773,13 @@ def do_send_messages(messages_maybe_none):
         Message.objects.bulk_create([message['message'] for message in messages])
         ums = []  # type: List[UserMessage]
         for message in messages:
-            # Outgoing webhook bots don't store UserMessage rows; they will be processed later.
```
Would anyone happen to know why we don't need to store a `UserMessage` row here?

robot-dreams commented:
By the way, one thing I forgot to mention:

I'm imagining that we could manually create a canonical dictionary of embedded bots that we'd want to support, where the keys are `bot_type` (strings) and the values are some sort of `BotHandler` object.

The `EmbeddedBotsWorker` can then use the `bot_type` to lookup the appropriate `BotHandler`, and then run the `BotHandler` on the message.

robot-dreams commented:
@timabbott Thanks for the suggestions yesterday!  Here's take 2.

vabs22 commented on zerver/models.py:
```diff
@@ -1738,18 +1758,40 @@ class CustomProfileFieldValue(models.Model):
+    # Valid interfaces are:
+    # - OUTGOING_WEBHOOK
+    # - EMBEDDED
+    interface = models.PositiveSmallIntegerField(default=OUTGOING_WEBHOOK)  # type: int
```
Interface here was for a separate thing initially. `Generic` class in [this](https://github.com/zulip/zulip/pull/3667/commits/f06c2448fc12377a700c352dedabfa0aef2582a4) commit is an example of interface. I don't know the current plan though. 
Also services can also be differentiated if they belong to embedded bot or outgoing webhook bot by `service.user_profile.is_embedded_bot`. If this differentiation is not required often, this method can be used. :)

robot-dreams replied:
Good call!  This is a much better idea :)

robot-dreams commented:
@timabbott I've cleaned this up and also fixed an issue with stripping leading mentions (both `@xkcd` and `@**xkcd**` should be stripped).  Let's try this out :)

timabbott commented:
Awesome, merged, thanks @robot-dreams!
